### ISSUE TITLE
Implement editing part for List Styles feature

### ISSUE THREAD

**Issue by pomek** [ckeditor/ckeditor5#7801]
Part of https://github.com/ckeditor/ckeditor5/issues/1031.

**Comment by pomek** [comment]
As for now, I have a basic implementation of the feature.

For a bulleted list, you can define a type of marker that will be used when rendering the list.

![image](https://user-images.githubusercontent.com/2270764/89767786-e8c31680-dafa-11ea-9108-e0b408cbcfdc.png)

The same applies for numbered lists:

![image](https://user-images.githubusercontent.com/2270764/89767808-f2e51500-dafa-11ea-973a-180aa1fb3a7b.png)

And even for nested and mixed lists:

![image](https://user-images.githubusercontent.com/2270764/89767828-ff696d80-dafa-11ea-94ba-c6cde7dbd2b6.png)


However, it's the beginning of problems since the feature must be integrated with other plugins.

### Indenting lists

When creating at some point a nested list, it should not inherit the list style attribute/value from the parent list.

```
■ List item 1.
■ List item 2.[]
■ List item 3.

editor.execute( 'indentList' );

■ List item 1.
    ○ List item 2.[]
■ List item 3.
```

However, the deeper structures should not be modified:

```
■ List item 1.
■ List item 2.[]
    ⁃ Nested item 1.
    ⁃ Nested item 2.
■ List item 3.

editor.execute( 'indentList' );

■ List item 1.
    ○ List item 2.[]
        ⁃ Nested item 1.
        ⁃ Nested item 2.
■ List item 3.
```

Also, it should work correctly when between list items, other nested lists are placed:

```
■ List item 1.
    ○ List item 2.
        ⁃ Nested item 1.
        ⁃ Nested item 2.
■ List item 3.[]
■ List item 4.

editor.execute( 'indentList' );

■ List item 1.
    ○ List item 2.
        ⁃ Nested item 1.
        ⁃ Nested item 2.
    ○ List item 3.[]
■ List item 4.
```

The attribute was restored to the default value. However, if `List item 2` has modified the list style, `List item 3` should copy the list style from `2`.

```
■ List item 1.
    ‣ List item 2.
        ⁃ Nested item 1.
        ⁃ Nested item 2.
■ List item 3.[]
■ List item 4.

editor.execute( 'indentList' );

■ List item 1.
    ♥ List item 2.
        ⁃ Nested item 1.
        ⁃ Nested item 2.
    ♥ List item 3.[]
■ List item 4.
```


### Outdenting lists

More or less it should revert changes made by the `indentList` feature.

A simple scenario:

```
■ List item 1.
    ○ List item 2.[]
■ List item 3.

editor.execute( 'outdentList' );

■ List item 1.
■ List item 2.[]
■ List item 3.
```

Don't modify deeper structures:

```
■ List item 1.
    ○ List item 2.[]
        ⁃ Nested item 1.
        ⁃ Nested item 2.
■ List item 3.

editor.execute( 'outdentList' );

■ List item 1.
■ List item 2.[]
    ⁃ Nested item 1.
    ⁃ Nested item 2.
■ List item 3.
```

And copy the value for the `listStyle` attribute from an item that is on the same (indent) level:

```
■ List item 1.
    ○ List item 2.
        ⁃ Nested item 1.
        ⁃ Nested item 2.
    ○ List item 3.[]
■ List item 4.

editor.execute( 'outdentList' );

■ List item 1.
    ○ List item 2.
        ⁃ Nested item 1.
        ⁃ Nested item 2.
■ List item 3.[]
■ List item 4.
```

However, the outdent feature can replace the `listItem` element with the `paragraph` element. No action is required in such a situation. 

### Merging a block to a list

A case:

```
Paragraph[]
■ List item 1.
■ List item 2.
■ List item 3.

editor.execute( 'bulletedList' )
```

The block item (`paragraph`) will be replaced with the `listItem` element. And the question is, should it copy the lists styles? In other words, should it be like:

```
* Paragraph[]
■ List item 1.
■ List item 2.
■ List item 3.
```

or 

```
■ Paragraph[]
■ List item 1.
■ List item 2.
■ List item 3.
```

The later seems to be correct. Fortunately, the first item on the list cannot be nested, so it shouldn't be too hard to implement.

In all scenarios, I was considering operations between the same type of lists (all nested lists are bulleted or numbered, but never both at the same time). What in case of mixed lists? As for now, I am treating them as separate elements.

**Comment by pomek** [comment]
Another case:

![image](https://user-images.githubusercontent.com/2270764/89777419-f03eeb80-db0b-11ea-8bb5-a841a4f75182.png)

What should happen after removing the paragraph? Should the list be merged and all rendered with `circle` marker or treat those as separate lists?

**Comment by pomek** [comment]
Another case:

![image](https://user-images.githubusercontent.com/2270764/89803112-b46b4c80-db32-11ea-9a97-8ebc2d46d482.png)

After outdenting, items 2.1 and 2.2 should inherit list styles from the list item 3.

**Comment by pomek** [comment]
TL;DR: Many cases I was able to cover on the model side. However, the conversion from the model to the view is a little buggy (the fourth issue).

---

In the beginning, I was thinking about introducing an additional attribute for the `listItem` element that would trigger the conversion and handling the list-style for lists. I went this way. Every `listItem` has the `listStyle` attribute with the `"default"` value if not specified. 

The value is being read from the list parent element (`ul`/`ol`) during [the upcast conversion](https://github.com/ckeditor/ckeditor5/blob/6f80796d02d7cca782214a1d215af8b6ec9bf92a/packages/ckeditor5-list/src/liststylesediting.js#L87-L94). Executing it with the low priority ensures that the element is always correct. If a list has not specified the `list-style-type` value, the default one will be used. It works.

But the solution isn't perfect. When you add a todo-list item, the attribute still is present in the model. I used a post-fixer to delete the attribute (`listStyle`) for all todo-list items.

When creating a bulleted/numbered list item then executing the `todoList` command, the model state will be correct. Unfortunately, it created an empty undo step. This is my first issue.

The value of the `list-style-type` style does not have to match the type of list that it's being used. I mean, you can use the unordered list (`<ul>`) with the `list-style-type: decimal` and it will use numbers as list markers.

The same operation you can do in the editor. It means if you used e.g unordered list with the square list-style then switched to a numbered list, the editor doesn't render any changes (only `listType` has changed for `listItems `elements).

The issue can be solved by restoring the default value for the `listStyle` attribute on changed items. 

A code snippet:

* At the end of the `LinkCommand#exexute()` method, emit an event:
	```js
	/**
	 * Event fired by the {@link #execute} method.
	 *
	 * It allows to execute an action after executing the {@link ~ListCommand#execute} method, e.g. adjusting
	 * attributes of changed blocks.
	 *
	 * @event executeCleanup
	 */
	this.fire( 'executeCleanup', blocks );
	```

* And attach listeners:
	```js
	this.listenTo( editor.commands.get( 'bulletedList' ), 'executeCleanup', restoreDefaultListStyle( editor ) );
	this.listenTo( editor.commands.get( 'numberedList' ), 'executeCleanup', restoreDefaultListStyle( editor ) );
	
	function restoreDefaultListStyle( editor ) {
		return ( evt, changedItems ) => {
			changedItems = changedItems.filter( item => item.is( 'element', 'listItem' ) );
	
			editor.model.change( writer => {
				for ( const item of changedItems ) {
					writer.removeAttribute( 'listStyle', item ); // setAttribute( 'listStyle', 'default', item );
				}
			} );
		};
	}
	```

It must be done in the same `model.change()` block that is used by `ListCommand#execute()` (one undo step for all operations). The same event-mechanism was used when indenting/outdenting the lists. This is the second issue since I am unsure about the solution.

We could use the `RemoveFormat` feature to remove any applied style for the list.

```js
editor.model.schema.setAttributeProperties( 'listStyle', {
	isFormatting: true
} );
```

However, the feature removes the attribute from the selection. If the selection is collapsed, only one list item will be changed. If non-collapsed, some items will be restored, some don't.

It won't be unified with the List style feature itself because you can apply the list style at any list item and the change will be applied to all items (that have the same indent, list type, and list style values).

At this point, I'm thinking the Remove format feature should not work with list styles (the third issue).

The last issue is about merging lists. Considering a structure like this:

```html
    <ul style="list-style-type: square;">
      <li>2</li>
      <li>3</li>
    </ul>
    <p>Paragraph</p>
    <ul style="list-style-type: disc;">
        <li>3</li>
        <li>4</li>
    </ul>
```

After deleting content from the paragraph, then remove the entire element, lists should not merge because they have different styles. The first idea @Reinmar had was [overwriting the check whether lists should be merged](https://github.com/ckeditor/ckeditor5/blob/6f80796d02d7cca782214a1d215af8b6ec9bf92a/packages/ckeditor5-list/src/utils.js#L178-L181). Unfortunately, it does not help.

While the first three issues aren't so urgent, they don't break the editor and everything still works and looks good, the last issue must be fixed before I'll start thinking about preparing the PR.

**Comment by pomek** [comment]
Maybe listening to `editor#deleteContent()` with higher priority could help. We could detect whether the list will be merged and adjust the `listStyle` attribute for the merged (second) list.

**Comment by pomek** [comment]
All issues I've spotted are extracted into a separate issue: https://github.com/ckeditor/ckeditor5/issues/7879
